RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo_gpu.cu

```diff
@@ -0,0 +1,1033 @@
+
+                    // consider flagging as kOutOfTime only if above noise
+                    if (uncalibRecHit.amplitude() > pedRMSVec[0] * amplitudeThreshEB){
+                        float outOfTimeThreshP = outOfTimeThreshG12pEB;
+                        float outOfTimeThreshM = outOfTimeThreshG12mEB;
+                        // determine if gain has switched away from gainId==1 (x12 gain)
+                        // and determine cuts (number of 'sigmas') to ose for kOutOfTime
+                        // >3k ADC is necessasry condition for gain switch to occur
+                        if (uncalibRecHit.amplitude() > 3000.) {
+                            for (int iSample = 0; iSample < EcalDataFrame::MAXSAMPLES; 
+                                 iSample++) {
+                                int GainId = edf.sample(iSample).gainId();
+                                if (GainId!=1) {
+                                    outOfTimeThreshP = outOfTimeThreshG61pEB;
+                                    outOfTimeThreshM = outOfTimeThreshG61mEB;
+                                    break;
+                                }
+                            } 
+                        }
+                        float correctedTime = (crh.timeMax-5) * clockToNsConstant + itimeconst + offsetTime;
+                        float cterm         = EBtimeConstantTerm;
+                        float sigmaped      = pedRMSVec[0];  // approx for lower gains
+                        float nterm         = EBtimeNconst*sigmaped/uncalibRecHit.amplitude();
+                        float sigmat        = std::sqrt( nterm*nterm  + cterm*cterm   );
+                        if ( ( correctedTime > sigmat*outOfTimeThreshP )   ||
+                             ( correctedTime < -sigmat*outOfTimeThreshM ))
+                        {
+                            uncalibRecHit.setFlagBit( EcalUncalibratedRecHit::kOutOfTime );
+                        }
+                    }
+                } // if not barrel
+            } // if timeAlgo_ ==
+        }
+    }
+}
+
+void scatter(host_data& h_data, device_data& d_data) {
+
+/*
+void scatter(EcalDigiCollection const& digis,
+             EcalUncalibratedRecHitCollection& rechits,
+             std::vector<EcalPedestal> const& vpedestals,
+             std::vector<EcalMGPAGainRatio> const& vgains,
+             std::vector<EcalXtalGroupId> const& vxtals,
+             std::vector<EcalPulseShape> const& vpulses,
+             std::vector<EcalPulseCovariance> const& vcovariances,
+             SampleMatrixGainArray const& noisecors,
+             device_data &d_data) {
+*/
+    auto const& ids = h_data.digis->ids();
+    auto const& digis_data = h_data.digis->data();
+    using digis_type = std::vector<uint16_t>;
+    using dids_type = std::vector<uint32_t>;
+    
+    //
+    // TODO: remove per event alloc/dealloc -> do once at the start
```

Comment:
Currently the explored model (even if not used everywhere yet) is to use a caching allocator to have a programming model of allocating device and pinned-host memory on every event without actually calling `cudaMalloc()` etc every time.

Whether to manage the device and host memories separately or use the unified memory is still under discussion. I suppose in the latter case we still want a caching allocator to avoid calling `cudaMallocManaged()` every time.